### PR TITLE
Move FetchPack functionality from NetworkOPs to LedgerMaster

### PR DESCRIPTION
**Pull request by nbougalis**
No new functionality.

Reviews: @vinniefalco, @seelabs

### CODE REVIEW

**Comment by nbougalis** [comment]
(will address the non-unity build failure)

File: src/ripple/app/ledger/impl/LedgerMaster.cpp [hunk 1]
```diff
@@ -959,144 +969,22 @@ class LedgerMasterImp
         return *ret;
     }
 
-    // Try to publish ledgers, acquire missing ledgers
-    void doAdvance ()
+    bool shouldFetchPack (std::uint32_t seq)
```
Comment (seelabs) [inline src/ripple/app/ledger/impl/LedgerMaster.cpp]: `shouldFetchPack` should probably be `const` and not change `fetch_seq_` Can we chage `fetch_seq_` in `getFetchPack` instead? If not then consider changing the name of this function to sound less like a query only function.

File: src/ripple/app/ledger/impl/LedgerMaster.cpp [hunk 2]
```diff
@@ -959,144 +969,22 @@ class LedgerMasterImp
...
+    }
 
-                    setPubLedger(ledger);
-                    progress = true;
-                }
+    bool shouldAcquire (
+        std::uint32_t const currentLedger,
+        std::uint32_t const ledgerHistory,
+        std::uint32_t const ledgerHistoryIndex,
+        std::uint32_t const candidateLedger);
```
Comment (seelabs) [inline src/ripple/app/ledger/impl/LedgerMaster.cpp]: `const`

File: src/ripple/app/ledger/impl/LedgerMaster.cpp [hunk 3]
```diff
@@ -1579,21 +1468,33 @@ class LedgerMasterImp
...
-{
-}
+    bool getFetchPack (
+        uint256 const& hash,
+        Blob& data) override;
+
+    void makeFetchPack (
+        Job&, std::weak_ptr<Peer> wPeer,
+        std::shared_ptr<protocol::TMGetObjectByHash> request,
+        uint256 haveLedgerHash,
```
Comment (seelabs) [inline src/ripple/app/ledger/impl/LedgerMaster.cpp]: Both wPeer and request could be `const&`

File: src/ripple/app/ledger/impl/LedgerMaster.cpp [hunk 4]
```diff
@@ -1608,11 +1509,341 @@ bool LedgerMaster::shouldAcquire (std::uint32_t const currentLedger,
...
+        auto msg = std::make_shared<Message> (reply, protocol::mtGET_OBJECTS);
+        peer->send (msg);
+    }
+    catch (...)
+    {
+        m_journal.warning << "Exception building fetch pach";
+    }
+}
+
+std::size_t LedgerMasterImp::getFetchPackCacheSize ()
```
Comment (seelabs) [inline src/ripple/app/ledger/impl/LedgerMaster.cpp]: Both `getFetchPackCacheSize` and `TaggedCache::getCacheSize` should be `const`.

**Comment by seelabs** [comment]
Nonunity build is broken. You can cherry pick this commit to add the missing includes: https://github.com/seelabs/rippled/commit/7efd10fbf2209da6842b39dfaed96f7aacbf8508

**Comment by nbougalis** [comment]
Nice, thanks @seelabs. I had fixed it last night, but neglected to push. I'll update.

**Comment by rec** [comment]
(Don't forget to add labels!  :-) )

File: src/ripple/app/ledger/LedgerMaster.h [hunk 5]
```diff
@@ -147,18 +153,44 @@ class LedgerMaster
...
+    std::size_t getFetchPackCacheSize () = 0;
 };
 
 std::unique_ptr <LedgerMaster>
-make_LedgerMaster (Config const& config, beast::Stoppable& parent,
-    beast::insight::Collector::ptr const& collector, beast::Journal journal);
+make_LedgerMaster (
+    Config const& config,
+    beast::Stoppable& parent,
+    beast::abstract_clock <std::chrono::steady_clock>& clock,
```
Comment (vinniefalco) [inline src/ripple/app/ledger/LedgerMaster.h]: Prefer `Stopwatch` `<ripple/basics/chrono.h>`

File: src/ripple/app/ledger/LedgerMaster.h [hunk 6]
```diff
@@ -24,14 +24,20 @@
 #include <ripple/app/ledger/LedgerHolder.h>
 #include <ripple/basics/StringUtilities.h>
 #include <ripple/protocol/RippleLedgerHash.h>
+#include <ripple/protocol/STValidation.h>
 #include <ripple/core/Config.h>
+#include <beast/chrono/basic_seconds_clock.h>
```
Comment (vinniefalco) [inline src/ripple/app/ledger/LedgerMaster.h]: `<ripple/basics/chrono.h>`

File: src/ripple/app/ledger/LedgerMaster.h [hunk 7]
```diff
@@ -147,18 +153,44 @@ class LedgerMaster
...
+
+    virtual
+    bool getFetchPack (
+        uint256 const& hash,
+        Blob& data) = 0;
+
+    virtual
+    void makeFetchPack (
+        Job&, std::weak_ptr<Peer> wPeer,
+        std::shared_ptr<protocol::TMGetObjectByHash> request,
```
Comment (vinniefalco) [inline src/ripple/app/ledger/LedgerMaster.h]: `const&`?

**Comment by vinniefalco** [comment]
I'd like to see a de-virtualization of `LedgerMaster`

**Comment by vinniefalco** [comment]
Ping?

**Comment by seelabs** [comment]
@nbougalis This may have fallen off your radar: nonunity build still fails

**Comment by nbougalis** [comment]
It hasn't - I'm on it.

**Comment by seelabs** [comment]
:+1:

**Comment by vinniefalco** [comment]
If it builds on msvc its good
:+1:

**Pull request closed without merging**